### PR TITLE
MRG: SSS Forward (for GSoC)

### PR DESCRIPTION
**Pull request by wronk**
This PR should eventually allow for forwards solutions to be computed directly from sensor-space to SSS space (as opposed to source space -> sensor space -> SSS space). In theory, this should reduce some noise associated with numerical imprecision and head movements.

My current goal is to add comments at every step so that the current forward operations can be linked back to white papers. I'll ask for suggestions on formatting and efficiency later, so please hold off on those critiques for the moment. If you notice any of the math description is incorrect though, please point it out -- the physics is not uncomplicated...

### CODE REVIEW

**Comment by wronk** [comment]
cc @Eric89GXL @staulu
Ongoing PR for computing forward solution to SSS space directly.

File: mne/forward/_compute_forward.py [hunk 1]
```diff
@@ -103,14 +104,16 @@ def _concatenate_coils(coils):
 
 def _bem_specify_coils(bem, coils, coord_frame, mults, n_jobs):
     """Set up for computing the solution at a set of coils"""
-    # Compute the weighting factors to obtain the magnetic field
-    # in the linear potential approximation
+    # Make sure coils are in MRI coordinate frame
```
Comment (Eric89GXL) [inline mne/forward/_compute_forward.py]: This is because the BEM is in MRI coords, right?

File: mne/forward/_compute_forward.py [hunk 2]
```diff
@@ -180,63 +195,129 @@ def _make_ctf_comp_coils(info, coils):
...
+    bem_rr : ndarray, shape (n_bem_rr, 3)
+        BEM source locations concatenated from all layers in MRI space.
+    n_jobs : int
+        Number of jobs to run in parallel
+    coil_type : str
+        'meg' or 'eeg'
+
+    Returns
+    -------
+    B
```
Comment (Eric89GXL) [inline mne/forward/_compute_forward.py]: This is the forward solution IIRC. Check the shape, should be (n_src * 3) x n_sensors or the reverse

**Comment by Eric89GXL** [comment]
BTW please keep this PR separate from one where you actually do meaningful work / changes if possible. So get the understanding first, changing variable names if necessary as you have done here, then we merge. Then we extend functionality. Otherwise the reviewing burden will be too high.

File: mne/forward/_compute_forward.py (hunk 1 above)
Comment (wronk) [inline mne/forward/_compute_forward.py]: Yep, I believe so. Updating comment.

**Comment by wronk** [comment]
Good plan, I'll focus on adding docstrings, linking to relevant papers, and variable names in this PR.

**Comment by wronk** [comment]
I sent an email to Matti about this, but if anyone has suggested materials for understanding the gory details of the forward solution calculation, please let me know. I'm currently working off of Hamalainen and Sarvas, 1989, and Mosher et al, 1999.

**Comment by mshamalainen** [comment]
Is this the idea?

Let y be the measurements, A_1 the innerspace multipole forward solution and x_1 the multipole coefficients. Then SSS approximates 

y = A_1 x_1

On the other hand, from dipoles we have

y = A x

The we can get a forward solution like matrix approximating x_1 in the least squares sense for give x from 

x_1 = (A_1^T A_1)^{-1} A_1^TA x

- Matti



> On Jun 4, 2015, at 5:47 AM, wronk <notifications@github.com> wrote:
> 
> I sent an email to Matti about this, but if anyone has suggested materials for understanding the gory details of the forward solution calculation, please let me know. I'm currently working off of Hamalainen and Sarvas, 1989, and Mosher et al, 1999.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/mne-tools/mne-python/pull/2167#issuecomment-108693508>.
> 



The information in this e-mail is intended only for the person to whom it is
addressed. If you believe this e-mail was sent to you in error and the e-mail
contains patient information, please contact the Partners Compliance HelpLine at
http://www.partners.org/complianceline . If the e-mail was sent to you in error
but does not contain patient information, please contact the sender and properly
dispose of the e-mail.

**Comment by mshamalainen** [comment]
To avoid the “double source modeling” induced by SSS + ordinary source modellling, I think the sober way to do this is the following

Let A_2 be the “outer space” multipole gain matrix, P_2 the corresponding orthogonal projection operator, and A the ordinary dipole gain matrix from the cortical sources to the sensors.

Then, apply P_2 to both y and A:

P_2 y = P_2 A x

and work with P_2 y and P_2 A instead of y and A. The reconstruction of the inner space multipole coefficients is quite unnecessary. Head movement compensation is, of course another issue which is not covered by this.

- Matti


> On Jun 4, 2015, at 5:47 AM, wronk <notifications@github.com> wrote:
> 
> I sent an email to Matti about this, but if anyone has suggested materials for understanding the gory details of the forward solution calculation, please let me know. I'm currently working off of Hamalainen and Sarvas, 1989, and Mosher et al, 1999.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/mne-tools/mne-python/pull/2167#issuecomment-108693508>.
> 



The information in this e-mail is intended only for the person to whom it is
addressed. If you believe this e-mail was sent to you in error and the e-mail
contains patient information, please contact the Partners Compliance HelpLine at
http://www.partners.org/complianceline . If the e-mail was sent to you in error
but does not contain patient information, please contact the sender and properly
dispose of the e-mail.

**Comment by wronk** [comment]
This is ready for heavy commenting/checking. This PR adds a fair amount documentation (with refs to relevant papers) and changes some variable names for clarity.

There 2-3 spots where pretty sure what's being calculated but not 100%. There are question marks next to those comments.

cc: @Eric89GXL @agramfort

**Comment by Eric89GXL** [comment]
You broke something

File: mne/forward/_compute_forward.py [hunk 3]
```diff
@@ -51,26 +59,77 @@ def _check_coil_frame(coils, coord_frame, bem):
...
-        # Accumulate the coefficients for each triangle node
-        # and add to the corresponding coefficient matrix
-        tri_rr = rr[tri]
+def _do_lin_field_coeff(bem_rr, tris, tn, ta, rmags, cosmags, ws, n_int):
+    """Compute field coefficients (parallel-friendly). See section IV of
+    Mosher et al., 1999 (specifically equation 35).
+
+    Parameters
+    ----------
+    bem_rr : ndarray, shape(n_BEM_vertices, 3)
```
Comment (agramfort) [inline mne/forward/_compute_forward.py]: missing space after shape

same pb below everywhere

File: mne/forward/_compute_forward.py [hunk 4]
```diff
@@ -459,46 +663,68 @@ def _magnetic_dipole_field_vec(rrs, coils):
 
 @verbose
 def _prep_field_computation(rr, bem, fwd_data, n_jobs, verbose=None):
-    """Precompute some things that are used for both MEG and EEG"""
-    cf = FIFF.FIFFV_COORD_HEAD
-    srr = mults = mri_Q = head_mri_t = None
+    """Precompute/store some things that are used for both MEG and EEG
+    (including multiplication factors, coordinate transforms, compensations,
+    and forward solutions) and store them in fwd_data.
```
Comment (agramfort) [inline mne/forward/_compute_forward.py]: pep257 says 1 line the multiple lines in docstring

**Comment by agramfort** [comment]
besides LGTM

File: mne/forward/_compute_forward.py [hunk 5]
```diff
@@ -126,22 +209,41 @@ def _bem_specify_coils(bem, coils, coord_frame, mults, n_jobs):
...
-    """Set up for computing the solution at a set of electrodes"""
+    """Set up for computing the solution at a set of EEG electrodes.
+
+    Parameters
+    ----------
+    bem : dict
+        BEM information
+    els : list of dict, len(n_EEG_sensors)
+        List of EEG sensor information dicts
+    mults: ndarray, shape (1, n_BEM_vertices)
```
Comment (Eric89GXL) [inline mne/forward/_compute_forward.py]: is this actually 2D? I think it's 1D which would be `shape (n_BEM_vertices,)`. this shows up a few places, so if it's wrong here please fix everywhere

**Comment by Eric89GXL** [comment]
thanks for the much improved docs @wronk, just check my last comment and then I'll merge

FYI once a PR reaches a state that you think it is done you can change the title to `MRG: `, it reminds us when looking at the list of PRs to take a look and make sure the person isn't waiting for comments

File: mne/forward/_compute_forward.py (hunk 5 above)
Comment (wronk) [inline mne/forward/_compute_forward.py]: It is actually 2D with that shape. Confirmed with breakpoints and looking at when it's instantiated (line 693).

**Comment by wronk** [comment]
Will do, thanks for the tip :)

File: mne/forward/_compute_forward.py (hunk 5 above)
Comment (Eric89GXL) [inline mne/forward/_compute_forward.py]: ahh, cool. I wonder why I did it that way...

**Comment by Eric89GXL** [comment]
Thanks @wronk! Now onto SSS, eh? :)

**Comment by wronk** [comment]
Yessir, working through the kinks on creating the actual multipolar basis set atm.

**Pull request merged**